### PR TITLE
Offir fsstore2

### PR DESCRIPTION
**Pull request by offirko**
### Description
Delivering FileSystemStore to KVStore Integration

### Pull request type
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### CODE REVIEW

**Comment by offirko** [comment]
@theamirocohen @davidsaada @dannybenor @yossi2le - FileSystemStore PR into KVStore Integration

File: features/storage/kvstore/filesystemstore/FileSystemStore.h [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+      *
+      * @param[in]  key                  Key.
+      * @param[in]  buffer               Value data buffer.
+      * @param[in]  buffer_size          Value data buffer size.
+      * @param[out] actual_size          Actual read size.
+      * @param[in]  offset               Offset to read from in data.
+      *
+      * @returns 0 on success or a negative error code on failure
+      */
+    virtual int get(const char *key, void *buffer, size_t buffer_size, size_t *actual_size, size_t offset = 0);
```
Comment (davidsaada) [inline features/storage/kvstore/filesystemstore/FileSystemStore.h]: The `actual_size` parameter should also be optional (if 0, you shouldn't return the value).

File: features/storage/kvstore/filesystemstore/FileSystemStore.cpp [hunk 2]
```diff
@@ -0,0 +1,613 @@
...
+{
+    int status = MBED_SUCCESS;
+    set_handle_t handle;
+
+    if (false == _is_initialized) {
+        status = MBED_ERROR_NOT_READY;
+        goto exit_point;
+    }
+
+    if ((!is_valid_key(key)) || (size > KVStore::MAX_KEY_SIZE) || ((buffer == NULL) && (size > 0)) ) {
```
Comment (davidsaada) [inline features/storage/kvstore/filesystemstore/FileSystemStore.cpp]: The size check here is wrong - you're checking the data size against the key size. This check should be removed (we don't have a limit for the data size).

File: features/storage/kvstore/filesystemstore/FileSystemStore.cpp [hunk 3]
```diff
@@ -0,0 +1,613 @@
...
+        status = MBED_ERROR_INVALID_SIZE;
```
Comment (davidsaada) [inline features/storage/kvstore/filesystemstore/FileSystemStore.cpp]: Should return invalid argument here.

File: features/storage/kvstore/filesystemstore/FileSystemStore.cpp [hunk 4]
```diff
@@ -0,0 +1,613 @@
...
+    _mutex.unlock();
+
+    return status;
+}
+
+int FileSystemStore::get_info(const char *key, info_t *info)
+{
+    int status = MBED_SUCCESS;
+    File kv_file;
+
```
Comment (davidsaada) [inline features/storage/kvstore/filesystemstore/FileSystemStore.cpp:238]: Should check key validity here.

File: features/storage/kvstore/filesystemstore/FileSystemStore.cpp [hunk 5]
```diff
@@ -0,0 +1,613 @@
...
+    int status = MBED_SUCCESS;
+    inc_set_handle_t *set_handle = NULL;
+    File kv_file;
+    key_metadata_t key_metadata;
+    int key_len = 0;
+
+    // Only a single key file can be incrementaly editted at a time
+    _mutex.lock();
+
+    if ((!is_valid_key(key)) || (handle == NULL) ) {
```
Comment (davidsaada) [inline features/storage/kvstore/filesystemstore/FileSystemStore.cpp]: Doesn't `verify_key_file` already perform the key validation?

File: features/storage/kvstore/filesystemstore/FileSystemStore.cpp [hunk 6]
```diff
@@ -0,0 +1,613 @@
...
+
+    return status;
+}
+
+int FileSystemStore::remove(const char *key)
+{
+    File kv_file;
+    key_metadata_t key_metadata;
+
+    if (key == NULL) {
```
Comment (davidsaada) [inline features/storage/kvstore/filesystemstore/FileSystemStore.cpp]: Isn't this checked by `verify_key_file`?

**Review by davidsaada — CHANGES_REQUESTED** [review]

File: features/storage/kvstore/filesystemstore/FileSystemStore.cpp (hunk 2 above)
Comment (offirko) [reply features/storage/kvstore/filesystemstore/FileSystemStore.cpp]: oops, will fix

File: features/storage/kvstore/filesystemstore/FileSystemStore.cpp (hunk 4 above)
Comment (offirko) [reply features/storage/kvstore/filesystemstore/FileSystemStore.cpp:238]: key validity is checked within _verify_key_file(), common to get(), get_info() and remove()

File: features/storage/kvstore/filesystemstore/FileSystemStore.cpp (hunk 6 above)
Comment (offirko) [reply features/storage/kvstore/filesystemstore/FileSystemStore.cpp]: you're right , now that I've added is_valid_key() its no longer needed - I'll remove.
Same below for set_start.

File: features/storage/kvstore/filesystemstore/FileSystemStore.h (hunk 1 above)
Comment (offirko) [reply features/storage/kvstore/filesystemstore/FileSystemStore.h]: Thanks.The code itself is implemented like you said. I'll fix the prototype.

**Comment by offirko** [comment]
@davidsaada - pushed updated version with fixes related to your CR

**Pull request closed without merging**